server/s3_put_multiobject_copy_action.cc

```diff
@@ -0,0 +1,962 @@
+}
+
+void S3PutMultipartCopyAction::copy_part_object_success() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+
+  s3_copy_part_action_state = S3CopyPartActionState::writeComplete;
+  object_data_copier.reset();
+  next();
+
+  s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+}
+
+void S3PutMultipartCopyAction::copy_part_object_failed() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+
+  s3_copy_part_action_state = S3CopyPartActionState::writeFailed;
+  set_s3_error(object_data_copier->get_s3_error());
+
+  object_data_copier.reset();
+  send_response_to_s3_client();
+
+  s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+}
+
+void S3PutMultipartCopyAction::save_metadata() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+  std::string s_md5_got = request->get_header_value("content-md5");
+  if (!s_md5_got.empty() && !motr_writer->content_md5_matches(s_md5_got)) {
+    s3_copy_part_action_state = S3CopyPartActionState::metadataSaveFailed;
+    s3_log(S3_LOG_ERROR, request_id, "Content MD5 mismatch\n");
+    set_s3_error("BadDigest");
+    send_response_to_s3_client();
+    return;
+  }
+  // Reset 'Date' and 'Last-Modified' time object metadata
+  part_metadata->reset_date_time_to_current();
+  part_metadata->set_content_length(std::to_string(total_data_to_stream));
+  part_metadata->set_md5(motr_writer->get_content_md5());
+  for (auto it : request->get_in_headers_copy()) {
+    if (it.first.find("x-amz-meta-") != std::string::npos) {
+      part_metadata->add_user_defined_attribute(it.first, it.second);
+    }
+  }
+  // bypass shutdown signal check for next task
+  check_shutdown_signal_for_next_task(false);
+
+  if (s3_fi_is_enabled("fail_save_part_mdata")) {
+    s3_fi_enable_once("motr_kv_put_fail");
+  }
+
+  part_metadata->save(
+      std::bind(&S3PutMultipartCopyAction::save_object_metadata_success, this),
+      std::bind(&S3PutMultipartCopyAction::save_metadata_failed, this));
+  s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+}
+
+void S3PutMultipartCopyAction::save_object_metadata_success() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+  s3_copy_part_action_state = S3CopyPartActionState::metadataSaved;
+  next();
+}
+
+void S3PutMultipartCopyAction::save_metadata_failed() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+  s3_copy_part_action_state = S3CopyPartActionState::metadataSaveFailed;
+  if (part_metadata->get_state() == S3PartMetadataState::failed_to_launch) {
+    s3_log(S3_LOG_ERROR, request_id,
+           "Save of Part metadata failed due to pre launch failure\n");
+    set_s3_error("ServiceUnavailable");
+  } else {
+    s3_log(S3_LOG_ERROR, request_id, "Save of Part metadata failed\n");
+    set_s3_error("InternalError");
+  }
+  send_response_to_s3_client();
+  s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+}
+
+const char xml_decl[] = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n";
```

Comment:
I see these declarations in S3CopyObjectAction class in middle too, ideally it should at the top of the file, after includes

Reply:
These declarations are no longer required.